File: bindings/android/conformance/src/main/kotlin/com/tacenta/conformance/Conversation.kt
// The conversation every check of the Kotlin head runs: two users sign up and
// in, find each other, message both ways, and one resumes from exported state
// and receives again. The same steps, line for line, as the TypeScript head's
// sdk/typescript/conformance/conversation.mjs and the Swift head's
// bindings/swift/conformance/Conversation.swift; tooling/check-conformance-lines.sh
// keeps the three saying the same thing. Kept apart from the JVM entry point
// so a device run can compile the same file.
package com.tacenta.conformance

import uniffi.tacenta_ffi.Address
import uniffi.tacenta_ffi.ClientException
import uniffi.tacenta_ffi.Message
import uniffi.tacenta_ffi.RestoreOutcome
import uniffi.tacenta_ffi.Tenant
import uniffi.tacenta_ffi.asFlow
import java.util.UUID
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.first

class Mismatch(message: String) : Exception(message)

private fun suffix() = UUID.randomUUID().toString().replace("-", "").take(8)

fun show(address: Address) = "${address.user}/${address.device}"

private fun expect(ok: Boolean, message: () -> String) {
    if (!ok) throw Mismatch(message())
}

/** The kind a call fails with, as the other heads spell it, or "accepted"
 *  if it did not fail. */
private suspend fun kindOf(call: suspend () -> Unit): String {
    return try {
        call()
        "accepted"
    } catch (e: ClientException) {
        // The subclass name, lowerCamel: SignInRefused -> signInRefused.
        val name = e::class.simpleName ?: "unknown"
        name.replaceFirstChar { it.lowercase() }
    } catch (e: Throwable) {
        "an untyped error: $e"
    }
}

private fun expectTexts(messages: List<Message>, expected: List<String>, what: String) {
    val got = messages.map { String(it.plaintext) }
    expect(got == expected) { "$what $got, expected $expected" }
}

/** Run the conversation on [tenant], reporting each step through [log].
 *  Throws on the first mismatch. */
suspend fun conversation(tenant: Tenant, log: (String) -> Unit) {
    val a = "conform-a-${suffix()}"
    val b = "conform-b-${suffix()}"
    val password = "p-${suffix()}-${suffix()}"
    tenant.signUp(a, password)
    tenant.signUp(b, password)
    log("signed up $a and $b")

    val alice = tenant.signIn(a, password)
    val bob = tenant.signIn(b, password)
    log("signed in as ${show(alice.address())} and ${show(bob.address())}")

    val refused = kindOf { tenant.signIn(a, "$password-wrong") }
    expect(refused == "signInRefused") { "a wrong password was refused as $refused" }
    log("a wrong password was refused as $refused")
    val taken = kindOf { tenant.signUp(a, password) }
    expect(taken == "usernameTaken") { "a second sign-up was refused as $taken" }
    log("a second sign-up was refused as $taken")

    val toBob = alice.find(b)?.address ?: throw Mismatch("find($b) gave null, expected ${show(bob.address())}")
    expect(toBob == bob.address()) { "find($b) gave ${show(toBob)}, expected ${show(bob.address())}" }
    expect(alice.find("nobody-${suffix()}") == null) { "find of a stranger was not null" }
    log("found $b at ${show(toBob)}")

    val nobody = kindOf {
        alice.send(Address("${toBob.user.substringBefore("/")}/nobody-${suffix()}", 1u), "to no one".toByteArray())
    }
    expect(nobody == "notFound") { "a send to an unregistered address was refused as $nobody" }
    log("a send to an unregistered address was refused as $nobody")

    alice.send(toBob, "conformance: first contact".toByteArray())
    val inbox = bob.receive()
    expectTexts(inbox, listOf("conformance: first contact"), "$b received")
    expect(inbox[0].from == alice.address()) { "from was ${show(inbox[0].from)}" }
    log("$b received the first message from ${show(inbox[0].from)}")

    bob.send(alice.address(), "conformance: reply".toByteArray())
    expectTexts(alice.receive(), listOf("conformance: reply"), "$a received")
    log("$a received the reply")

    // A receive left pending does not hold the client: a send goes through
    // meanwhile, and the pending receive gets the reply it provokes.
    coroutineScope {
        val pending = async { alice.receive() }
        alice.send(toBob, "conformance: while receiving".toByteArray())
        expectTexts(bob.receive(), listOf("conformance: while receiving"), "$b received while $a was receiving")
        bob.send(alice.address(), "conformance: to the pending receive".toByteArray())
        expectTexts(pending.await(), listOf("conformance: to the pending receive"), "the pending receive of $a")
    }
    log("$a sent while its own receive was pending, and that receive got the reply")

    // The stream form: the same receive underneath, one message at a time.
    bob.send(alice.address(), "conformance: through the stream".toByteArray())
    val streamed = alice.inbound().asFlow().first()
    expectTexts(listOf(streamed), listOf("conformance: through the stream"), "the inbound stream of $a")
    log("$a took the next message from its inbound stream")

    val state = alice.exportState()
    val again = tenant.signInWithState(a, password, state)
    val outcome = again.restoreOutcome()
    expect(outcome == RestoreOutcome.RESUMED) { "the restore outcome of $a was $outcome" }
    bob.send(again.address(), "conformance: after resume".toByteArray())
    expectTexts(again.receive(), listOf("conformance: after resume"), "resumed $a received")
    log("$a resumed from ${state.size} bytes of state and received again")
}

File: bindings/swift/conformance/Conversation.swift
// The conversation every check of the Swift head runs: two users sign up and
// in, find each other, message both ways, and one resumes from exported state
// and receives again. The same steps, line for line, as the TypeScript head's
// sdk/typescript/conformance/conversation.mjs and the Kotlin head's
// bindings/android/conformance Conversation.kt; tooling/check-conformance-lines.sh
// keeps the three saying the same thing.
import Foundation
import Tacenta

struct Mismatch: Error, CustomStringConvertible { let description: String }

private func suffix() -> String { String(UInt32.random(in: 0 ..< 0xFFFF_FFFF), radix: 16) }

func show(_ address: Address) -> String { "\(address.user)/\(address.device)" }

private func expect(_ ok: Bool, _ message: @autoclosure () -> String) throws {
    if !ok { throw Mismatch(description: message()) }
}

/// The kind a call fails with, as the other heads spell it, or "accepted"
/// if it did not fail.
private func kindOf(_ call: () async throws -> Void) async -> String {
    do {
        try await call()
        return "accepted"
    } catch let error as ClientError {
        // The case name, lowerCamel: `SignInRefused(reason: ...)` -> signInRefused.
        let name = String(describing: error).split(separator: "(")[0]
        return name.prefix(1).lowercased() + name.dropFirst()
    } catch {
        return "an untyped error: \(error)"
    }
}

private func expectTexts(_ messages: [Message], _ expected: [String], _ what: String) throws {
    let got = messages.map { String(decoding: $0.plaintext, as: UTF8.self) }
    try expect(got == expected, "\(what) \(got), expected \(expected)")
}

/// Run the conversation on `tenant`, reporting each step through `log`.
/// Throws on the first mismatch.
func conversation(_ tenant: Tenant, _ log: (String) -> Void) async throws {
    let a = "conform-a-\(suffix())"
    let b = "conform-b-\(suffix())"
    let password = "p-\(suffix())-\(suffix())"
    try await tenant.signUp(username: a, password: password)
    try await tenant.signUp(username: b, password: password)
    log("signed up \(a) and \(b)")

    let alice = try await tenant.signIn(username: a, password: password)
    let bob = try await tenant.signIn(username: b, password: password)
    log("signed in as \(show(alice.address())) and \(show(bob.address()))")

    let refused = await kindOf { _ = try await tenant.signIn(username: a, password: "\(password)-wrong") }
    try expect(refused == "signInRefused", "a wrong password was refused as \(refused)")
    log("a wrong password was refused as \(refused)")
    let taken = await kindOf { try await tenant.signUp(username: a, password: password) }
    try expect(taken == "usernameTaken", "a second sign-up was refused as \(taken)")
    log("a second sign-up was refused as \(taken)")

    guard let toBob = try await alice.find(username: b)?.address else {
        throw Mismatch(description: "find(\(b)) gave nil, expected \(show(bob.address()))")
    }
    try expect(toBob == bob.address(), "find(\(b)) gave \(show(toBob)), expected \(show(bob.address()))")
    try expect(try await alice.find(username: "nobody-\(suffix())") == nil, "find of a stranger was not nil")
    log("found \(b) at \(show(toBob))")

    let nobody = await kindOf {
        try await alice.send(to: Address(user: "\(toBob.user.split(separator: "/")[0])/nobody-\(suffix())", device: 1),
                       message: Data("to no one".utf8))
    }
    try expect(nobody == "notFound", "a send to an unregistered address was refused as \(nobody)")
    log("a send to an unregistered address was refused as \(nobody)")

    try await alice.send(to: toBob, message: Data("conformance: first contact".utf8))
    let inbox = try await bob.receive()
    try expectTexts(inbox, ["conformance: first contact"], "\(b) received")
    try expect(inbox[0].from == alice.address(), "from was \(show(inbox[0].from))")
    log("\(b) received the first message from \(show(inbox[0].from))")

    try await bob.send(to: alice.address(), message: Data("conformance: reply".utf8))
    try expectTexts(try await alice.receive(), ["conformance: reply"], "\(a) received")
    log("\(a) received the reply")

    // A receive left pending does not hold the client: a send goes through
    // meanwhile, and the pending receive gets the reply it provokes.
    async let pending = alice.receive()
    try await alice.send(to: toBob, message: Data("conformance: while receiving".utf8))
    try expectTexts(try await bob.receive(), ["conformance: while receiving"], "\(b) received while \(a) was receiving")
    try await bob.send(to: alice.address(), message: Data("conformance: to the pending receive".utf8))
    try expectTexts(try await pending, ["conformance: to the pending receive"], "the pending receive of \(a)")
    log("\(a) sent while its own receive was pending, and that receive got the reply")

    // The stream form: the same receive underneath, one message at a time.
    try await bob.send(to: alice.address(), message: Data("conformance: through the stream".utf8))
    var streamed: [Message] = []
    for try await message in alice.inbound() {
        streamed.append(message)
        break
    }
    try expectTexts(streamed, ["conformance: through the stream"], "the inbound stream of \(a)")
    log("\(a) took the next message from its inbound stream")

    let state = try await alice.exportState()
    let again = try await tenant.signInWithState(username: a, password: password, state: state)
    let outcome = try await again.restoreOutcome()
    try expect(outcome == .resumed, "the restore outcome of \(a) was \(outcome)")
    try await bob.send(to: again.address(), message: Data("conformance: after resume".utf8))
    try expectTexts(try await again.receive(), ["conformance: after resume"], "resumed \(a) received")
    log("\(a) resumed from \(state.count) bytes of state and received again")
}
